### ISSUE TITLE
[proposal] Binary Payload Support

### ISSUE THREAD

**Issue by stealthycoin** [awslabs/chalice#348]
## Overview
This feature will add support for round-tripping binary data to a chalice deployed app. This includes correctly configuring API Gateway to handle the binary content types and correctly interacting with the API Gateway interface from Lambda. Ideally this would mostly be a transparent change for users, only requiring minor changes in some cases.

Some `Content-Types` will be treated as binary data automatically and will require no change in code to be properly handled by API Gateway.
Default content types that are considered binary:
```
application/epub+zip
application/octet-stream
application/ogg
application/pdf
application/vnd.amazon.ebook
application/x-7z-compressed
application/x-bzip
application/x-bzip2
application/x-rar-compressed
application/x-tar
application/zip
audio/aac
audio/midi
audio/ogg
audio/webm
audio/x-wav
font/ttf
font/woff
font/woff2
image/bmp
image/gif
image/jpeg
image/png
image/svg+xml
image/tiff
image/webp
image/x-icon
video/3gpp
video/3gpp2
video/mpeg
video/ogg
video/webm
```

## Motivation
Open issues related to binary payloads:
#223 
#250  

## Interface
The interface will be very minimal since this change can mostly be taken care of by the framework automatically. In the case where someone wants to send a binary payload and the content type they are using (or forced to use if the request is coming from a third party they do not have control over) is not in this default list, they can customize the list by calling an app level function (`app.api.define_binary_types`) to add more types. 
Example of adding a custom binary type `application/custom-type`:
```python
from chalice import Chalice
app = Chalice('custom-binary')
app.api.define_binary_types(['application/x-special-binary-type'])
```

## Implementation
API Gateway takes requests and matches their `Content-Type` header against a user provided list per restapi resource. Chalice will provide this list through the swagger file property `x-amazon-apigateway-binary-media-types`. When API Gateway transforms this request to lambda event, it will base64 encode the body, and add a top level key called `isBase64Encoded` with a value of `True`. Chalice will check for the presence of the `isBase64Encoded=True` in the event, and base64decode the body before forwarding it on to the corresponding route handler.

For responses, Chalice will check the headers dictionary for a `Content-Type` key with a value matching a known binary type, using the same list that was provided to API Gateway. If the type matches it will base64 encode the body, and attach the `isBase64Encoded=True` key/value to the dictionary it sends to API Gateway. 

## Example
Below is a full chalice application that echos back any binary content POSTed to it. All you would have to do to accept binary data is add the appropriate content type to the route, and be sure to set a binary type `Content-Type` header in the response.
```python
    demo = Chalice('demo-app')

    @demo.route('/bin-echo', methods=['POST'], 
                content_types=["application/octet-stream"])
    def bin_echo():
        raw_body = demo.current_request.raw_body
        return app.Response(
            status_code=200,
            body=raw_body,
            headers={'Content-Type': "application/octet-stream"})

```

**Comment by kyleknap** [comment]
Looks good. I think this approach is good under the assumption a user will only ever create one chalice application. If a user can create more than one application, there may be issues where if they try to add more binary types that will be shared across chalice applications. Here is some other feedback:

* May be good to rename ``BINARY_TYPES`` to something else that is specific to API Gateway. Maybe like ``BINARY_CONTENT_TYPES``?
* It would be good to be more detailed on how you are going to enable binary types using API Gateway's API and how that all works for completeness.

**Comment by stealthycoin** [comment]
I think may be even `APIGATEWAY_BINARY_CONTENT_TYPES` to be super clear may be useful. I was trying to keep the implementation details out of the proposal, but I can update it with some details.

**Comment by kyleknap** [comment]
I don't think we need the API Gateway part of it. I think that is too specific to implementation details because this is mainly concerned with properly handling binary content of HTTP requests (ignoring the API gateway bit). I also do not think the implementation details of how you enable it needs to be super long (maybe 3 or 4 sentences). It would be just nice to have a little more background on what will be happening when the binary type is set for a route or is returned back.

**Comment by jamesls** [comment]
In general looks good.  A few notes:

* Image processing seems to be a popular use case.  I'd add a few more image types to the default list (`image/png`, etc).
* -1 to global state modification.  This will also make testing awkward.  App level scoping should be as "global" as we allow.
* I would actually prefer we scope it even further.  As @kyleknap mentioned this is really just specific to the HTTP part so even making it that more delineated could be useful.  What about something like:  `chalice.api.define_binary_types('application/foo', 'application/bar', ...)`  This also starts the processing of partition API gateway specific things into a separate child object of an app.

**Comment by stealthycoin** [comment]
I like this idea: `chalice.api.define_binary_types(...)` this is what I wanted to do initially but I didn't think it fit well with what we already had. To be clear that would still be global state though, or would this be off of the app object itself? I would prefer `app.api.define_binary_types`.

**Comment by jamesls** [comment]
So given the limitation on the max number of binary types what if we had:

1. Have a default list of built in binary types.
2. Expose this default list as a public attr somewhere
3. Give the user an API to completely override this list entirely.

So from your referenced PR we could just make `DEFAULT_BINARY_TYPES` global and then just let the override:

```
# Overide
app.api.define_binary_types(['application/foo', 'application/bar', 'application/baz', ...])

# Add to existing
app.api.define_binary_types(app.api.DEFAULT_BINARY_TYPES + ['application/another',
 'application/and-another'])

What do you all think?

**Comment by kyleknap** [comment]
So the `app.api.DEFAULT_BINARY_TYPES` you cannot extend directly from `app.api.DEFAULT_BINARY_TYPES`?

What was the reason for not having a `app.api.BINARY_TYPES` that you can extend or replace outright? So something like:
```
# Overide
app.api.BINARY_TYPES = ['application/foo', 'application/bar', 'application/baz', ...]

# Add to existing
app.api.BINARY_TYPES.extend(['application/another', 'application/and-another'])
```

**Comment by jamesls** [comment]
In the linked PR, the `BINARY_TYPES` is a class attribute, which is global.  I'm not in favor of an API that encourages modifying global state.

Perhaps instead of `BINARY_TYPES`, there's just a `default_binary_types` property that returns a copy each time to make it even harder to mutate the list of default values?

**Comment by kyleknap** [comment]
Yeah that's better to have a `default_binary_types` property. I thought it was a weird that I would be interacting with a global class attribute `DEFAULT_BINARY_TYPES` and a `define_binary_types` when I would just be able to extend `DEFAULT_BINARY_TYPES` anyways.

**Comment by stealthycoin** [comment]
Merged in with #352
